### PR TITLE
Negative axis support for gradient of reduce_prod

### LINKED ISSUE #10835: Gradient of reduce_prod does not support negative axis
The gradient of `reduce_prod` does not support negative axis unlike `reduce_prod` itself.
It is apparently caused by `gather` not supporting negative axes.
This code illustrates the problem.
```python
import tensorflow as tf

vars = tf.Variable([[1., 2.], [3., 4.]])
prod = tf.reduce_prod(vars, -1) # Negative axis here

tf.InteractiveSession()
tf.global_variables_initializer().run()
print(prod.eval()) # Works fine
print(tf.gradients(prod, vars)[0].eval()) # Crashes
```

### PR DESCRIPTION
**Pull request by AnishShah**
Fixes  #10835

### CODE REVIEW

**Comment by AnishShah** [comment]
@tensorflow-jenkins bot didn't get triggered. Is there any problem?

**Comment by drpngx** [comment]
Jenkins, test this please.

**Review by drpngx — APPROVED** [review]

**Comment by AnishShah** [comment]
PyLint tests were failing. So, I fixed that. Can you run Jenkins again?

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by AnishShah** [comment]
All tests have passed now! Is this ready to merge?

**Comment by drpngx** [comment]
Yep. Congrats!
